src/transformers/models/ict/modeling_ict.py

```diff
@@ -0,0 +1,810 @@
+                    def custom_forward(*inputs):
+                        return module(*inputs, output_attentions)
+
+                    return custom_forward
+
+                layer_outputs = torch.utils.checkpoint.checkpoint(
+                    create_custom_forward(block),
+                    hidden_states,
+                )
+            else:
+                layer_outputs = block(hidden_states, output_attentions)
+
+            hidden_states = layer_outputs[0]
+
+            if output_attentions:
+                all_self_attentions = all_self_attentions + (layer_outputs[1],)
+
+        if not return_dict:
+            return tuple(v for v in [hidden_states, all_hidden_states, all_self_attentions] if v is not None)
+
+        return BaseModelOutput(
+            last_hidden_state=hidden_states,
+            hidden_states=all_hidden_states,
+            attentions=all_self_attentions,
+        )
+
+
+class ResnetBlock(nn.Module):
+    def __init__(self, dim):
+        super().__init__()
+        self.conv_block = nn.Sequential(
+            nn.ReflectionPad2d(2),
+            nn.utils.spectral_norm(
+                nn.Conv2d(in_channels=dim, out_channels=dim, kernel_size=3, padding=0, dilation=2, bias=False)
+            ),
+            nn.ReLU(True),
+            nn.ReflectionPad2d(1),
+            nn.utils.spectral_norm(
+                nn.Conv2d(in_channels=dim, out_channels=dim, kernel_size=3, padding=0, dilation=1, bias=False)
+            ),
+        )
+
+    def forward(self, x):
+        out = x + self.conv_block(x)
+        # Remove ReLU at the end of the residual block
+        # http://torch.ch/blog/2016/02/04/resnets.html
+
+        return out
+
+class InpaintGenerator(nn.Module):
+    def __init__(self, residual_blocks=8):
+        super().__init__()
+
+        self.encoder = nn.Sequential(
+            nn.ReflectionPad2d(3),
+            nn.Conv2d(in_channels=6, out_channels=64, kernel_size=7, padding=0),
+            nn.ReLU(True),
+            nn.Conv2d(in_channels=64, out_channels=128, kernel_size=4, stride=2, padding=1),
+            nn.ReLU(True),
+            nn.Conv2d(in_channels=128, out_channels=256, kernel_size=4, stride=2, padding=1),
+            nn.ReLU(True),
+        )
+
+        blocks = [ResnetBlock(256) for _ in range(residual_blocks)]
+
+        self.middle = nn.Sequential(*blocks)
+
+        self.decoder = nn.Sequential(
+            nn.ConvTranspose2d(in_channels=256, out_channels=128, kernel_size=4, stride=2, padding=1),
+            nn.ReLU(True),
+            nn.ConvTranspose2d(in_channels=128, out_channels=64, kernel_size=4, stride=2, padding=1),
+            nn.ReLU(True),
+            nn.ReflectionPad2d(3),
+            nn.Conv2d(in_channels=64, out_channels=3, kernel_size=7, padding=0),
+        )
+
+    def forward(self, x):
+        x = self.encoder(x)
+        x = self.middle(x)
+        x = self.decoder(x)
+        x = (torch.tanh(x) + 1) / 2
+
+        return x
```

Comment:
@alaradirik As we discussed earlier, this is now encoder, middle, decoder. But I noticed that in the original code, `InpaintGenerator_5` (which is same as `InpaintGenerator` above) is instantiated [in `InpaintingModel`](https://github.com/raywzy/ICT/blob/289c7dd86fc67d9726abed703b241676ff260d8d/Guided_Upsample/src/models.py#L64). 

And then, its `forward` function looks like [this](https://github.com/raywzy/ICT/blob/289c7dd86fc67d9726abed703b241676ff260d8d/Guided_Upsample/src/models.py#L165-L183) (to simply since `self.config.Generator==0`):
```python
def forward(self, images, edges, masks):
    images_masked = (images * (1 - masks).float()) + masks
    inputs = torch.cat((images_masked, edges), dim=1)
    outputs = self.generator(inputs)
    return outputs
```

Reply:
So what I did was create just put `InpaintGenerator` in the init function of `ICTGuidedUpsampler` similar to the original code (in which case, `InpaintGenerator` is instantiated inside `InpaintModel`). And after deleting all else, it's looks pretty [barebone like this](https://github.com/huggingface/transformers/blob/05c1b553c143045c8c360bc9a8214ef1e28d9bd0/src/transformers/models/ict/modeling_ict.py#L653-L665). 

Or do you think it's just better to remove `InpaintGenerator` and just put everything inside `ICTGuidedUpsampler` (which is more simliar to what we discussed)?

Reply:
Also, `Discriminator`, `AdversarialLoss`, `StyleLoss`, `PerceptualLoss` is not used in this case (since those for training, I guess). Should we delete them or should we add them so that users can train it? (In that case, `Discriminator` will also be instantiated in `ICTGuidedUpsampler`, I assume?)